KAFKA-8199: Fix ClassCastException when trying to groupBy after suppress

ouertani opened the pull request:
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

Fixes #8199 ClassCastException when trying to groupBy after suppress

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

https://github.com/apache/kafka/blob/da53141d931342563f5bbcf6b4846671d59e3033/streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorTest.java#L451
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ X] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ouertani committed:
Fixes #8199 ClassCastException when trying to groupBy after suppress

ouertani committed:
Fixes #8199 ClassCastException when trying to groupBy after suppress

dejan2609 committed:
MINOR: Update Gradle to 5.4.1 and update its plugins  (#6436)

Details:
 * gradle: 5.1.1  -->  5.4.1
 * grgit: 1.9.3  -->  3.1.1 (breaking change release: artifact name is changed;
also, Grgit.open' usage is slightly refactored)
 * gradle-versions-plugin: 0.20.0 --> 0.21.0
 * shadow: 4.0.3  -->  4.0.4
 * spotless-plugin-gradle: 3.17.0  --> 3.23.0
 * checkstyle: 8.10 --> 8.20
 * spotbugs: 3.1.8 --> 3.1.12
 * jacoco: 0.8.2 --> 0.8.3

huxi-2b committed:
KAFKA-7779; Avoid unnecessary loop iteration in leastLoadedNode (#6081)

In NetworkClient.leastLoadedNode, we invoke `isReady` to  check if an established connection exists for the given node. `isReady` checks whether metadata needs to be updated also which wants to make metadata request first priority. However, if the to-be-sent request is metadata request, then we do not have to check this otherwise the loop in `leastLoadedNode` will do a complete iteration until the final node is selected. 

Reviewers: Jason Gustafson <jason@confluent.io>

dhruvilshah3 committed:
KAFKA-8134: `linger.ms` must be a long

Reviewers: Ismael Juma <ismael@juma.me.uk>, Colin P. McCabe <cmccabe@apache.org>

chasewalden committed:
MINOR: Fix Embedded ConfigDef Validator toString issue (#6339)

`ConfigDef.embeddedValidator` should return an Anonymous Object instead of lambda so that we can have a useful `toString()` for methods such as `toRst`.

Reviewers: Jason Gustafson <jason@confluent.io>

mjsax committed:
KAFAK-3522: add API to create timestamped stores (#6601)

Reviewers: A. Sophie Blee-Goldman <sophie@confluent.io>, Bruno Cadonna <bruno@confluent.io>, Guozhang Wang <guozhang@confluent.io>

mjsax committed:
KAFKA-3729: Revert adding Serde auto-config (#6630)

* Revert "MINOR: Add unit test for SerDe auto-configuration (#6610)"

This reverts commit 172fbb2dd55144e2e44777174f970b56768e1777.

* Revert "[KAFKA-3729] Auto-configure non-default SerDes passed alongside the topology builder (#6461)"

This reverts commit e56ebbffca57741d398283e46073ed4170f8f927.

The two merged PRs introduce a breaking change. Reverting to preserve backward compatibility. Jira ticket reopened.

Reviewers: Ted Yu <yuzhihong@gmail.com>, Guozhang Wang <guozhang@confluent.io>

stanislavkozlovski committed:
KAFKA-7992: Introduce start-time-ms metric (#6318)

Reviewers: Colin P. McCabe <cmccabe@apache.org>, Ismael Juma <ismael@juma.me.uk>

bbejeck committed:
KAFKA-8298: Fix possible concurrent modification exception (#6643)

When processing multiple key-changing operations during the optimization phase a ConcurrentModificationException is possible.

Reviewers: Guozhang Wang <wangguoz@gmail.com>

guozhangwang commented:
@ouertani could you rebase this PR?

@bbejeck @vvcephei could you take a look?

hzxa21 committed:
KAFKA-8066; Always close the sensors in Selector.close() (#6402)

When shutting down the ReplicaFetcher thread, we may fail to unregister sensors in selector.close(). When that happened, we will fail to start up the ReplicaFetcherThread with the same fetch id again because of the IllegalArgumentException in sensor registration. This issue will cause constant URPs in the cluster because the ReplicaFetchterThread is gone.

This patch addresses this issue by introducing a try-finally block in selector.close() so that we will always unregister the sensors in shutting down ReplicaFetcherThreads.

Reviewers: Rajini Sivaram <rajinisivaram@googlemail.com>, Jason Gustafson <jason@confluent.io>

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSuppressSupplier.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    public void enableSendingOldValues() {
+    }
```
maybe throw an unsupported operation exception here?

ouertani replied:
add  throw new UnsupportedOperationException("enableSendingOldValues is not supported by KTableSuppressSupplier");

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSuppressSupplier.java:
```diff
@@ -0,0 +1,77 @@
+    public Processor<K, Change<V>> get() {
+        return suppressProcessor;
+    }
+    private class KTableSuppressValueGetter implements KTableValueGetter<K, V> {
```
nit: missing empty line

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -401,14 +401,14 @@ public String queryableStoreName() {
                 keySerde,
                 valSerde == null ? null : new FullChangeSerde<>(valSerde)
-            );
+        );
```
nit: reduce the indention of the passed in parameters by 4, too.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorTest.java:
```diff
@@ -445,6 +448,20 @@ public void suppressShouldShutDownWhenOverByteCapacity() {
+        final Properties properties = new Properties();
+
+        builder.<String, String>stream("topic")
+                .groupByKey().windowedBy(TimeWindows.of(Duration.ofSeconds(30))).count()
```
nit: move `windowBy()` to it's own line -- also `count()` (similar below)

dejan2609 committed:
MINOR: Extract Gradle and its plugins versions into dependencies.gradle (#6648)

Reviewers: Ismael Juma <ismael@juma.me.uk>

ouertani committed:
Merge branch 'KAFKA-8199' of https://github.com/ouertani/kafka into KAFKA-8199

ouertani commented:
> @ouertani could you rebase this PR?
> 
> @bbejeck @vvcephei could you take a look?

@guozhangwang rebased.

ouertani committed:
minor format changes #8199

bbejeck commented:
Hi @ouertani looks like there are some extra changes not related to this PR. Can you try and do a force push of the rebased branch?

vvcephei commented:
Looking at the commits, it looks like it might have been a merge. The thing to do is `rebase -i trunk`, and then remove all the commits that aren't for this change.

ouertani commented:
@bbejeck @vvcephei supersedes https://github.com/apache/kafka/pull/6684
